src/main/java/za/co/neslotech/spring/trainspringbootinsight/controller/CrudController.java

```diff
@@ -0,0 +1,75 @@
+package za.co.neslotech.spring.trainspringbootinsight.controller;
+
+import jakarta.persistence.EntityNotFoundException;
+import jakarta.servlet.http.HttpServletRequest;
+import org.springframework.beans.BeanUtils;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.servlet.support.ServletUriComponentsBuilder;
+import za.co.neslotech.spring.trainspringbootinsight.service.CrudService;
+
+import java.util.List;
+
+/**
+ * A generic REST API Controller with CRUD operations.
+ * The child class should use the following annotations.
+ *      1. @RestController
+ * @param <T> The entity class.
+ * @param <I> The primary key field type.
+ */
+public abstract class CrudController<T, I> {
+
+    protected abstract CrudService<T, I> getService();
+
+    @GetMapping
+    public ResponseEntity<List<T>> list(){
+        return ResponseEntity.ok(getService().findAll());
+    }
+
+    @GetMapping
+    @RequestMapping("{id}")
+    public ResponseEntity<T> get(final @PathVariable I id){
+            return ResponseEntity.ok(getService().findById(id)
+                    .orElseThrow(() ->
+                            new EntityNotFoundException(String.format("Entity with id %s not found!", id))
+                    ));
```

Comment:
Not a big deal but you can let the service handle this scenario, the controller would then use the exception handler to handle this exception.